### ISSUE TITLE
Strategy for maintaining the thin and other hubs

### ISSUE THREAD

**Issue by ansell** [AtlasOfLivingAustralia/biocache-hubs#204]
The strategy for maintaining the thin hubs is not currently working. They are never receiving attention from web/grails developers during sprints, and I was only originally put on to redeploy them to a single AWS server once, not to maintain, upgrade, and implement missing functionality for them as has occurred.

The last interaction that @M-Nicholls would like me to have with the thin hubs (APA/ASBP/Tepapa/OZCAM/OBIS) is to coordinate an immediate next release (depends on #203), redeploy them once to AWS, and have Matt switch over the DNS so that they all point to AWS servers. After that they would need to be maintained by a grails developer.

The documentation for redeploying them to AWS has been added to the README.md files in each of the hubs and they all have ansible scripts setup by me on the last two maintenance sprints that I have been part of, but there is a continuous time requirement for upgrading them along with biocache-hubs (which is why this issue is located in this tracker).

After that, there needs to be thought given by grails developers about the most efficient way for them to maintain the hubs (in the time they have available between sprints or join with biocache sprints or create new sprints) that we have committed to maintaining, and remove the rest. The current method of overriding all of the CSS/JS/GSP and configuration as a grails project is difficult to maintain. 

APPD/AVH/MDBA are full hubs in their own right at this point and can't be considered thin hubs anymore. However, the rest are still relatively thin hubs consisting essentially only of skins and a few configuration values and there needs to be a new strategy for maintaining them efficiently.

APPD/AVH/MDBA hubs have been so highly customised that they will also require dedicated sprint time to maintain to relieve the burden that the Data Management Team, @sadeghim, @m-hope and I, have been taking on so far with performing grails maintenance that the web developers have not had time to perform.

CC @nickdos @djtfmartin @adam-collins @m-hope @M-Nicholls

**Comment by nickdos** [comment]
> The strategy for maintaining the thin hubs is not currently working.

We're not actually at the "maintaining" stage for hubs. This work was to get to that point - the currently deployed hubs are all running Spring MVC. So its not surprising that getting to this point has required a bit of pain. WE've put off doing this last painful bit of work for too long and that has been a mistake - it should've been done over a year ago.

Anyway, I'm on maintenance for the next sprint so I'll spend a bit of time fixing some of the issues (common ones across all hubs). I don't think it will be a major problem moving forward, if we keep them up to date. I think once they are deployed, then keeping them up to date should be minimal work - ansible is a big help in this regard.

APPD & MDBA are out of scope for this - they are, as you said quite different from the others - so we can ignore it for the purposes of this issue.

AVH is relatively up to date and is under ansible for deployment, so as far as I'm concerned it is no different to the other hubs in terms of general maintenance. 

I don't agree about the overriding of files being difficult to maintain, as I've been maintaining AVH with its differences for almost 2 years and the amount of work has been minimal. Getting the initial version working is difficult but maintaining it once that is done, is not difficult. So I think you're confusing those separate aspects.

**Comment by ansell** [comment]
From my perspective, the amount of time required to update dependencies on each hub to verify that the latest version of biocache-hubs and other plugins still work was taking me more time than I have in a three week maintenance sprint.

The dependency updates need to happen regularly to pickup changes made in biocache-hubs that make the thin hubs continue to with the latest versions of BIE/biocache.ala.org.au/lists/etc. I don't see the current versions of the thin hubs going into a time-efficient maintenance state without some changes to the way they are implemented.

The reason for bringing it up now is that @M-Nicholls has expressed to me that it is taking too much of my time when it goes outside of a sprint block as it has so far.

After the first maintenance sprint, all of the ansible scripts were created and the version bumps were done to get them all up to biocache-hubs-0.75, but after doing that part for the first time I didn't have time to complete functionality testing against them all. The testing that I had worked through showed issues that needed to be fixed in a future sprint before the DNS switch over was able to occur. 

In the second maintenance sprint that I just completed, if there were no incompatible changes to biocache-hubs, then I had planned on working on the grails configuration and missing GSP page issues that I created in the respective issue trackers during the previous sprint, with the goal of getting all of them over to aws-hubs.ala.org.au by the end of the three week period (except AVH which was deployed to a separate AWS server between my two sprints). 

However, there were incompatible changes made to biocache-hubs in either 1.0 or 1.1 between the two sprints (by no-ones fault, they were simply a result of adding new functionality to biocache-hubs) that meant I spent most of my time trying to diagnose and fix grails plugin dependency issues that were unrelated to the goal of deploying them and testing their functionality so that the DNS switch could occur by the end of last week when I was supposed to resume Data Management duties.

I don't think I am confusing the total time issues, although the availability of ansible scripts reduces the time in future for round-trip release and deploy cycles. Each of the 7 or so thin hubs (omitting MDBA/APPD for this discussion) contains customised Grails markup and groovy/java code that needs to be verified and possibly updated when plugin versions or Grails versions change. The plugins that were relevant in the last sprint were the auth, bootstrap, images, downloads and biocache-hubs plugins, which have non-trivial dependencies across them, and in some cases were released while still having -SNAPSHOT dependencies which made it difficult to verify some of the issues across different thin hubs built on Travis at different times. 

You were able to do the update and maintenance for AVH in your spare time, but there was no attention given to the other 6 between my two maintenance sprints so they lagged far enough behind that you couldn't clearly remember what changes you made to AVH so that I could make similar changes to the others in my 3 week allocation.

In the longer term, the update of Grails to a recent maintained version is yet to be placed in the Biocache roadmap (https://github.com/AtlasOfLivingAustralia/documentation/wiki/RoadMap), but it will require non-trivial work on each of the hubs that the ALA are committed to providing at that point. If we have eliminated custom Grails code from the thin hubs at that point (my preferred strategy for resolving this issue) and they are simply configuration driven, it would significantly reduce the amount of effort required for that step as if the plugins are functionally tested to work once, they will work in all 6/7/etc. cases without the duplication of time testing combinations with each one as I did during the sprint.

The references to Bootstrap 3 and Bootstrap 4 that are already in the Biocache roadmap also represent non-trivial amounts of effort for maintenance of the thin hubs, as they will all need to have their Bootstrap dependencies updated soon after biocache-hubs gets its update. That step can't really be avoided using configuration, as the skinning, that makes up the thin hubs core reason for existing, is all done using Bootstrap at some level. Each of thin hubs will need to be either updated or abandoned at that point even if we have configuration working to eliminate the custom Grails code.
